### PR TITLE
WIP: Descriptor updates

### PR DESCRIPTION
**Pull request by jgriffiths**
Adds more introspection information and support for multi-path descriptors.

This allows iteration/generation of all available possible scripts/addresses without knowing anything about a descriptor up-front.

cc: @Sjors @k-matsuzawa

### CODE REVIEW

**Review by instagibbs — COMMENTED** [review]
some light code review

File: src/test/test_descriptor.py [hunk 1]
```diff
@@ -195,6 +199,31 @@ def test_canonicalize_checksum_bad_args(self):
                ret, out = fn(descriptor, flags)
                self.assertEqual((ret, out), (WALLY_EINVAL, None))
 
+    def test_feautures(self):
```
[the remark this answers is not in the document]
Comment (jgriffiths) [reply src/test/test_descriptor.py]: Fixed in bbcedd07e699b2df7d1d1df1987216050b6a51d4, thanks.

File: src/test/test_bip32.py [hunk 2]
```diff
@@ -542,12 +546,21 @@ def get_paths(path):
...
+                 ('/1/<0;1>',     0, 0),          # Multi-index present without flag
+                 ('/1/<0;1>',     M, 0)]          # Multi-index flag present
 
         for path, flags, wildcard in cases:
             flags = flags | FLAG_KEY_PRIVATE
-            ret = bip32_key_from_parent_path_str_n(m, path, len(path), wildcard, flags, key_out)
+            ret = bip32_key_from_parent_path_str_n(m, path, len(path or ''),
+                                                   wildcard, flags, key_out)
             self.assertEqual(ret, WALLY_EINVAL)
+            if flags == M:
```
[the remark this answers is not in the document]
Comment (jgriffiths) [reply src/test/test_bip32.py:559]: `bip32_key_from_parent_path_str` doesn't have a `multi_index` argument and so can't be used with multi index paths. So this test just verifies that passing a MI path with or without the flag fails as expected for that call.

To use a multi path you need to convert the path to a regular path using `bip32_path_from_str` . Testing that requires the `multi_index` argument be passed, which is done in a separate case below this test so there isn't a redundant `multi_index` littering the  `bip32_key_from_parent_path_str` test cases.

File: src/bip32.c [hunk 3]
```diff
@@ -112,9 +112,31 @@ int bip32_path_from_str_n(const char *str, size_t str_len,
...
     while (i < str_len) {
-        bool is_wildcard = false;
+        size_t multi;
+        bool is_wildcard = false, is_multi = false;
         start = i;
         v = 0;
+        if (str[i] == '<' && (flags & BIP32_FLAG_STR_MULTIPATH)) {
+            /* Multi-path expression - skip to the required multi_index */
+            ++i;
+            found_multi = true;
```
[the remark this answers is not in the document]
Comment (jgriffiths) [reply src/bip32.c]: Yeah it turned out to be redundant, but its too painful to edit it out and rebase over!

**Pull request merged**